tests/test_directives.py

```diff
@@ -0,0 +1,266 @@
+# Copyright (c) 2017 The Regents of the University of Michigan
+# All rights reserved.
+# This software is licensed under the BSD 3-Clause License.
+import pytest
+import sys
+
+from flow.directives import Directives, _DirectivesItem
+from flow.directives import NP, NRANKS, NGPU, EXECUTABLE, OMP_NUM_THREADS
+from flow.directives import WALLTIME, MEMORY, PROCESS_FRACTION
+from flow.directives import no_aggregation
+from flow import FlowProject
+from tempfile import TemporaryDirectory
+
+
+@pytest.fixture()
+def available_directives_list():
+    return [NP, NRANKS, NGPU, OMP_NUM_THREADS,
+            EXECUTABLE, WALLTIME, MEMORY,
+            PROCESS_FRACTION]
+
+
+@pytest.fixture()
+def product_directive():
+    def val(v):
+        if v < 10:
+            raise ValueError("Price cannot be less than 10 units")
+        return v
+
+    def finalize(value, dict):
+        discount = dict.get('discount', 0)
+        free = dict.get('free', False)
+        value = value - discount
+        if value < 0 or free:
+            return 0
+        return value
+
+    product = _DirectivesItem(name='product', validation=val,
+                              default=10, serial=no_aggregation,
+                              parallel=no_aggregation, finalize=finalize)
+    return product
+
+
+@pytest.fixture()
+def non_default_directive_values():
+    def _values(values=0):
+        if values == 1:
+            return {'np': 4, 'ngpu': 1, 'nranks': 0,
+                    'omp_num_threads': 10, 'executable': 'PathFinder',
+                    'walltime': 20., 'memory': 16, 'processor_fraction': 0.5}
+        return {'np': 1, 'ngpu': 10, 'nranks': 5,
+                'omp_num_threads': 20, 'executable': 'Non Default Path',
+                'walltime': 64., 'memory': 32, 'processor_fraction': 0.5}
+    return _values
+
+
+class TestItems:
+    """
+    Tests for _DirectivesItem class
+    """
+    def test_default(self):
+        assert NP.default == 1
+        assert NGPU.default == 0
+        assert NRANKS.default == 0
+        assert OMP_NUM_THREADS.default == 0
+        assert EXECUTABLE.default == sys.executable
+        assert WALLTIME.default == 12.0
+        assert MEMORY.default == 4
+        assert PROCESS_FRACTION.default == 1.
+
+    def test_invalid_values(self):
+        invalid_values = {'np': [-1, 1., 'foo', {}, None],
+                          'ngpu': [-1, 1., 'foo', {}, None],
+                          'nranks': [-1, 1., 'foo', {}, None],
+                          'omp_num_threads': [-1, 1., 'foo', {}, None],
+                          'executable': [1, {}, None],
+                          'walltime': [-1., 'foo', {}, None],
+                          'memory': [-1, 4.5, 'foo', {}, None],
+                          'processor_fraction': [-0.5, 2.5, 'foo', {}, None]
+                          }
+        for dir_name in invalid_values:
+            for dir_value in invalid_values[dir_name]:
+                with pytest.raises(Exception):
```

Comment:
Can we be more specific on the type of exception? If we can't that is okay, but if we can that would be helpful. It may be worth ensuring we raise a `ValueError` for these.

Reply:
We can but since `ValueError` is not the only error which will get thrown. Hence I would have to define all the errors manually.

Reply:
In Python except blocks allow for you to pass a tuple like `except (ValueError, TypeError)`. `pytest.raises` allows a similar API for catching one of a number of exceptions.

Reply:
Oh alright, that's interesting